force-app/classes/src/MethodSpy.cls

```diff
@@ -18,32 +18,63 @@
 @IsTest
 public class MethodSpy {
   public String methodName;
-  private Integer callCount = 0;
-  private List<Object> lastCallParams;
+  private CallLog callLog;
   private List<ParameterizedMethodSpyCall> parameterizedMethodCalls = new List<ParameterizedMethodSpyCall>();
   private Object returnValue;
   private Boolean configuredGlobalReturn = false;
   private Exception exceptionToThrow;
 
   public MethodSpy(String methodName) {
     this.methodName = methodName;
+    this.callLog = new CallLog();
   }
 
-  public List<Object> getLastCallParams() {
-    return this.lastCallParams;
+  public Boolean hasBeenCalled() {
+    return !this.callLog.isEmpty();
   }
 
-  public Boolean hasBeenCalled() {
-    return this.callCount > 0;
+  public Boolean hasBeenCalledWith(List<Object> params) {
+    Iterator<CallParameter> it = this.callLog.iterator();
+    while (it.hasNext()) {
+      if (params == it.next().params) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public Boolean hasBeenCalledWith(ParametersMatcher matcher) {
+    Iterator<CallParameter> it = this.callLog.iterator();
+    while (it.hasNext()) {
+      if (matcher.matches(it.next().params)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  public Boolean hasBeenLastCalledWith(List<Object> params) {
+    if (this.callLog.isEmpty()) {
+      return false;
+    }
+    return this.callLog.getLast() == params;
```

Comment:
suggestion: introduce a static instance of `ParametersMatcher` that relies on `==` (maybe named `ParametersMatcher.Equals` and simplify all code that now do a `==` by calling the alternative that use a parameter matcher?